### DISCUSSION TITLE
REV 2025 Beta

### DISCUSSION THREAD

**Discussion by jfabellera** [wpilibsuite/2025Beta#27] [Announcements]
[See the latest announcement here](https://github.com/wpilibsuite/2025Beta/discussions/27#discussioncomment-11611738)

### Original Post
Apologies for the long wait! REV's 2025 beta is finally here. We look forward to hearing your feedback.

* [REVLib 2025.0.0-beta-1](https://github.com/REVrobotics/REV-Software-Binaries/releases/tag/revlib-2025.0.0-beta-1)
* [Instructions for updating Hardware Client and firmware](https://github.com/wpilibsuite/2025Beta/blob/main/REV.md)

For information on how to migrate to 2025 REVLib, please see [our documentation](https://docs.revrobotics.com/brushless/revlib/revlib-overview/migrating-to-revlib-2025)

> [!WARNING]
> After updating your SPARK MAX or Flex to 25.0.0, you will be unable to communicate with them through the Hardware Client if using a Power Distribution Hub or Pneumatic Hub as the USB CAN bridge. A later update will be available for the PDH and PH that fixes this. For now, connect to the Hardware Client via a SPARK MAX or Flex.

**Comment by person4268** [comment]
Setting some parameters in sim (signals, followers) results in a crash with 2025.0.0-beta-1:
```
Error at com.revrobotics.spark.SparkBase.configure(SparkBase.java:276): Unhandled exception instantiating robot com.revrobotics.spark.SparkBase java.lang.IllegalStateException: Parameter type mismatch for parameter id
        at com.revrobotics.spark.SparkBase.configure(SparkBase.java:276)
        at frc.robot.Robot.<init>(Robot.java:39)
        at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:301)
        at edu.wpi.first.wpilibj.RobotBase.lambda$startRobot$0(RobotBase.java:418)
        at java.base/java.lang.Thread.run(Thread.java:840)
```

I sent some more details to support@revrobotics but I can send them here or whereever else is needed. Just figured I'd post the high-level cause somewhere so other people can see it.

**Reply by jfabellera** [reply]
Same issue as #28 

This will be fixed in the next release which should be coming soon

**Reply by person4268** [reply]
Oh :facepalm:, that got posted after I finished debugging but before I wrote it up. Thanks you for fixing it.

**Reply by jfabellera** [reply]
Fix is out. See https://github.com/wpilibsuite/2025Beta/discussions/27#discussioncomment-11194292

**Comment by jfabellera** [comment]
REVLib beta 2 and SPARK MAX pre-release 6 have been released

* [REVLib 2025.0.0-beta-2](https://github.com/REVrobotics/REV-Software-Binaries/releases/tag/revlib-2025.0.0-beta-2)
* [Instructions for updating Hardware Client and firmware](https://github.com/wpilibsuite/2025Beta/blob/main/REV.md)

> [!WARNING]
> After updating your SPARK MAX or Flex to 25.0.0, you will be unable to communicate with them through the Hardware Client if using a Power Distribution Hub or Pneumatic Hub as the USB CAN bridge. A later update will be available for the PDH and PH that fixes this. For now, connect to the Hardware Client via a SPARK MAX or Flex.

### Issues fixed
- #32
- #28

**Comment by pigrammer3** [comment]
Does `SparkMax.configure` (and `SparkFlex.configure`) wait to verify that the configuration is applied? That is, is it a blocking call or does it run in the background? Do we need to wait some time for the configuration to save after calling the method, or can we use the motor immediately?

**Reply by jfabellera** [reply]
Yes it does verify that setting the parameters was successful (and retries if necessary). It is a blocking call, so once that's done, you're able to use the configured controller if no errors or exceptions come up.

**Reply by viggy96** [reply]
How many retries are attempted, and that number configurable? 

And is the total amount of time to wait for the operation to complete configurable?

**Reply by jfabellera** [reply]
> How many retries are attempted, and that number configurable?

It retries a maximum of 5 times and can be configured with [`setCANMaxRetries()`](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparklowlevel#setCANMaxRetries(int)) (I recognize that the docs will need to be updated for the new system but it's still inherently correct)

> And is the total amount of time to wait for the operation to complete configurable?

What operation are you talking about? All of `configure()` or individual parameter set commands? There isn't one for `configure()` as that would be directly related to the number of parameters being set from the configuration. Individual blocking calls that require a response (like the parameter set calls that `configure()` make) have timeouts that can be configured with [`setCANTimeout()`](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparkbase#setCANTimeout(int)). See my other comment about this method: https://github.com/wpilibsuite/2025Beta/discussions/27#discussioncomment-11243999

**Reply by viggy96** [reply]
I guess its taken care of via `setCANMaxRetries()`, and the timeout on each one attempt determines the maximum amount of time `configure()` will take.

**Comment by person4268** [comment]
If I want to toggle a motor from brake to coast mode, would it be faster (and semantically correct) to create a new SparkFlexConfig with just idlemode set to the correct value?

**Reply by jfabellera** [reply]
Yes and you would pass `kNoResetSafeParameters` to preserve the rest of the configuration and preferably `kNoPersistParameters` since that operation takes a considerable amount of time especially for the SPARK MAX, and blocking for any amount of time during robot operation should generally be avoided.

**Reply by NoahStricker** [reply]
Is there no longer a non-blocking way to set idle mode?

**Reply by jfabellera** [reply]
Not at the moment. The effect of setting the CAN timeout to 0 (as documented [here](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparkbase#setCANTimeout(int))) is no longer present due to some infrastructure changes. The ability to do certain commands asynchronously will be added back before kickoff.

**Reply by viggy96** [reply]
Would it be too much to ask for some wrapper methods for the more mundane tasks, such as setting idle mode, and inverting the motor?

**Comment by bryceroethel** [comment]
Great to see these improvements, thank you to the team that worked on this! Just a few feature requests/nitpicks:

- It appears that you can no longer get the fault/warning bit field without shims, which is a regression from last season. Is there a chance this could be added back, possibly just as another field of the Fault/Warning classes? It was useful for logging as it could easily be packed into a WPILib struct.

- MAXMotion looks really promising, however I am weary of the lack of feedforward types other than kV. For some applications, the current interface is mostly sufficient (static kG for elevators, and kS for any mechanism, is easy enough to implement on the RIO and pass via arbFF) however mechanisms like arms could benefit from an on-controller cosine based kG, and kA may also be useful in some cases. Are there any plans to implement more on-controller feedforward gains, or will kV remain as the only available option for this season?

- Last season, absolute encoder velocity by default would return rotations per second, unlike the motor's internal encoder which would return RPM (this inconsistency was undocumented). It appears that the documentation for the absolute encoder's config has been updated to now specify that the default base unit is RPM, however I didn't see anything in the change log specifying this fix. I haven't gotten the chance to test this yet on a real robot, was the documentation change an oversight that occurred while migrating the configuration API, or was a fix actually implemented?

**Reply by viggy96** [reply]
I'd say it would also be kinda weird in the code to use the WPILib feed forward for the other values, but REV feed forward for kV.

**Reply by jfabellera** [reply]
Thank you for the feedback!
* I definitely see how that can be useful, and we can add a bit field for that. We've had some reports that the short being returned by `getFaults()` wasn't very usable for some, so that's why we moved to returning a struct of human readable faults.
* I can't make any promises about adding additional feed forward gains this season, but it has definitely been on the list.
* The original documented RPS was actually a mistake in documentation. The native units returned by firmware was RPM, and it is still RPM. No changes to firmware have been made for this, only documentation.

**Reply by viggy96** [reply]
What is the recommended way to do feed forward then for various mechanisms? Assuming MAXMotion is doing motion profiling, having the controller do feed forward calculations is kinda necessary, as the feed forward is based on the setpoint, and a motion profile usually works by continually changing the setpoint. So the controller should be able to calculate not only the correct setpoint to apply given the motion profile at that point in time, but also the correct feed forward for that setpoint.

**Reply by narmstro2020** [reply]
Throwing another voice in the mix.  
Doing a mix of RIO feeforward math and controller feedback math is not really ideal.  You can only add the arbFF every 20 ms from the RIO while the feedback calculations are running faster than that.  

Question:  Can the controllers support ks, kg, kv, and ka with a cosine component for arm?
It seems like kV exists as kFF (which is really confusing and should be renamed to kV like every other usage in FRC).  

Also if I'm not mistaken kFF doesn't quite act the same way as WPILIB's kv and either needs to behave more like WPI or the usage (especially with SmartMotion) needs to be documented.

**Reply by narmstro2020** [reply]
And I don't want to beat a dead horse on this topic, but the intended use for set Reference is to set when you need to set.  Sure you could do constant sets and most likely will, but if you have a separate RIO led feedforward then you have to bury it in a periodic loop anyways.  Where as integrated it into setReference would keep the same set it and forget it nature (though of course at the millisecond level

**Comment by viggy96** [comment]
Is there an easy way to query the Spark for a config setting? I don't see any getters to that effect. 

Also, I've previously had a "SparkHelpers" class that had an identical package to REV's SparkBase so that I could create certain getters that weren't available via normal API, as well as some other methods that weren't expose either. Some examples include, getting smart current limit values (current, chop cycles, RPM), enabling/disabling center aligned mode on the absolute encoder (this allows the sensor to report -0.5 to +0.5 instead of 0.0 to 1.0), and converting the Spark fault bits to a human readable string.

https://github.com/lasarobotics/PurpleLib/blob/master/src/main/java/com/revrobotics/SparkHelpers.java

Are all those methods now exposed via the normal API?

**Reply by bryceroethel** [reply]
> Is there an easy way to query the Spark for a config setting? I don't see any getters to that effect.

- [Spark MAX](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparkmax#configAccessor)
- [Spark Flex](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparkflex#configAccessor)

**Reply by jfabellera** [reply]
I believe the only one from the list that is not in the API is support for the center aligned mode for absolute encoder. There were reasons why we didn't include it in the past, but I believe we should be able to add that now. Let me know if I am missing anything.

**Reply by viggy96** [reply]
What about the following method?
```
  /**
   * Enable mode which sets the output of the PID controllers to be voltage instead of duty cycle.
   *
   * <p>To disable, change or disable voltage compensation. Those settings will overwrite this one
   */
  public static REVLibError enablePIDVoltageOutput(CANSparkBase spark) {
    return REVLibError.fromInt(CANSparkMaxJNI.c_SparkMax_SetParameterUint32(spark.sparkMaxHandle, 74, 1));
  }
  ```

**Reply by NoahStricker** [reply]
Is the using the config accessor a blocking call?

**Reply by jfabellera** [reply]
> Is the using the config accessor a blocking call?

Yes, as documented in the [Javadoc](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparkmax#configAccessor)

**Comment by viggy96** [comment]
Is there any methodology to tuning MAXMotionPosition/MAXMotionVelocity that teams should be aware of? Will a high level explainer of REV's implementation of this algorithm be released?

**Reply by jfabellera** [reply]
A docs page for MAXMotion will be added before kickoff and will include a tuning guide and best practices.

**Comment by viggy96** [comment]
How do I change the status frame period in the new API?

**Reply by jasondaming** [reply]
It’s part of the [signals config](https://codedocs.revrobotics.com/java/com/revrobotics/spark/config/sparkbaseconfig#signals). Some signals are always enabled (status frame 0) and most are disabled by default. Calling any getter will automatically enable the associated signal, so in most cases no manual configuration is required. The config object lets you force enable any signal and customize periods.

Shamelessly stolen from @jwbonner when he answered my question about the same thing.

**Comment by viggy96** [comment]
What are the default measurement period and average depth for the internal motor encoder for a NEO Vortex and NEO 1.1?

I believe the optimal values that we previously found were 32ms/8 for a Spark Flex, and 16ms/2 for Spark Max.

**Reply by jfabellera** [reply]
The defaults are documented in the [EncoderConfig javadocs](https://codedocs.revrobotics.com/java/com/revrobotics/spark/config/encoderconfig#quadratureAverageDepth(int)).

One thing to note is that the NEO Vortex uses the quadrature set of parameters and the NEO 1.1 uses the UVW set of parameters.

**Reply by viggy96** [reply]
Can these defaults be changed to reflect the more optimal values? 

As it stands these settings have to be changed in order to achieve better velocity PID performance, especially on the NEO v1.1.

And less experienced teams won't know to do that.

**Reply by viggy96** [reply]
Those values I specified were found by former REV employee Will Toth by the way.

**Reply by viggy96** [reply]
Has there been any thought to changing the defaults for this setting? 

The existing defaults cause the measurements to lag, negatively impacting velocity PID performance.

**Comment by viggy96** [comment]
Will the API provide detailed feedback on which portion of a config failed to apply if any?

**Reply by jfabellera** [reply]
Errors including parameter set failures should be logged to the driver station with the parameter that failed.

**Comment by viggy96** [comment]
Is there a `toString()` method for a `Faults` object?

**Reply by jfabellera** [reply]
Not at the moment.

**Comment by viggy96** [comment]
Is there a method to automatically reset the motor encoder or external encoder to a user specified value upon connected limit switch activation? 

Right now we can technically do this by creating a Trigger, but it would be best if this functionality was in the Spark firmware itself and its internal control loop.

**Reply by jfabellera** [reply]
Not at the moment. We can add this to our backlog.

**Comment by viggy96** [comment]
Can `SparkModel` and `getSparkModel()` be made public?

**Reply by jfabellera** [reply]
We will add it to the list.

**Comment by viggy96** [comment]
Is there a method to limit supply current?

**Reply by jfabellera** [reply]
There is no input current limiting at the moment.

**Comment by Liam-Stow** [comment]
Is the spark simulation expected to work in C++ yet? I have tried the bare-bones test below and this does not change the position or velocity of the spark in sim gui at all. The equivalent code works in Java.

```cpp
void Robot::TeleopInit() {
  _sparkFlex.Set(1);
}

void Robot::SimulationPeriodic() {
  double rpm = 100;
  double batteryVoltage = 12;
  double deltaTime = 0.02;
  _sparkFlexSim.iterate(rpm, batteryVoltage, deltaTime);
}
```

**Reply by Liam-Stow** [reply]
Looks like this is still the behavior in REVLib beta 3, I've made an issue here: #55

**Comment by jfabellera** [comment]
REVLib beta 3 and SPARK MAX and Flex pre-release 7 have been released

* [REVLib 2025.0.0-beta-3](https://github.com/REVrobotics/REV-Software-Binaries/releases/tag/revlib-2025.0.0-beta-3)
* [Instructions for updating Hardware Client and firmware](https://github.com/wpilibsuite/2025Beta/blob/main/REV.md)

> [!WARNING]
> After updating your SPARK MAX or Flex to 25.0.0, you will be unable to communicate with them through the Hardware Client if using a Power Distribution Hub or Pneumatic Hub as the USB CAN bridge. A later update will be available for the PDH and PH that fixes this. For now, connect to the Hardware Client via a SPARK MAX or Flex.

### Issues fixed
- #44 
- #43 
- #41

**Comment by levyishai** [comment]
Is there any real reason to make the SparkSim accept a DCMotor in its constructor? 
If the entire idea of the SparkSim is that it allows for an external physics sim, why not go all the way with it?
As far as I can see, the only place the DCMotor is used is for determining the current. I think a more elegant and nice solution would be to add a function to set the current, and adding the current as a variable in `iterate`. This removes redundancy, shows the idea of SparkSim better, and allows for other ways to simulate current (i.e. current spikes at certain situations).

**Reply by narmstro2020** [reply]
It would be better to just have accept methods that take in the value from the DCMotor.  This would allow support from the other Sims as well.

**Reply by narmstro2020** [reply]
In fact It would be better to have constructors for ElevatorSim and SingleJointedArmSim and create Consumers and Suppliers for internal use for whichever Sim class that you use in your constructor.  It looks like the only reason DCMotorSim is referenced in the SparkSim is give it access to the DCMotorSim's current getter.  So a DoubleSupplier would work for that.  

Or take out the DCMotorSim argument from the constructor and pass in the current into the iterate method along with velocity.

**Reply by narmstro2020** [reply]
So I should really read better.  I forgot it took in DCmotor not DCMotorSim.  Ignore the previous 2 comments.  

Any of the sim classes can produce their own value for Current.

**Comment by thenetworkgrinch** [comment]
Are there any plans to add a new `ArbFFUnit` for Amperage that way we can take advantage of new feedforwards on the SparkMax/SparkFlex?

**Comment by narmstro2020** [comment]
Units class support would be nice :)

**Comment by JochemvdBroek** [comment]
I have a question regarding the motion profile generation for MAXMotion. Specifically, are motion profiles generated based on the current position of the motor, and are the they recreated if the motor is not at its goal position after following a motion profile? 

Previous years we have used the WPILib TrapeziodProfile, and this only uses the currently sampled setpoint of the profile to generate the next one (as far as I have seen in the documentation). It does not take into account the current position of the motor and instead relies of position PID feedback to correct for positional errors. The main advantage of this, is that the motion profiles are relatively 'clean' and no additional profiles are generated if the system is not at its setpoint fast enough. When we were testing MAXMotion we saw some behaviour that indicated that motion profiles were regenerated based on the current position of the motor and we wanted to know if this is the case.

**Reply by jfabellera** [reply]
Yes, the profile is live re-calculated with the current position of the motor.

> are the they recreated if the motor is not at its goal position after following a motion profile?

I want to say yes, but I'm not entirely sure what you are asking. If the profile is constantly being re-calculated, then it will always be trying to reach the setpoint. If the motor moves outside the allowed error, it will try to move back towards the setpoint.

**Reply by viggy96** [reply]
@jfabellera What is the recommended way to handle feed forward when using MAXMotionPosition? 

Ideally this would be handled internally, since feed forward values are dependent on setpoint.

**Reply by JochemvdBroek** [reply]
> but I'm not entirely sure what you are asking.

I was indeed wondering if the profile was constantly being re-calculated based on the current position. What benefits does this bring compared to only using the current setpoint to generate the motion profile? If I recall correctly, the WPILib implementation of the motion profile only uses the current setpoint and not the actual position to generate the next setpoint and this has worked great for us so far. 

Another reason why I am asking this, is that we saw some unexpected behavior during our first tests and we thought that some of the issues we encountered might be caused by the fact that the profile is based on the current position. We thought that because the position was not behaving as expected because of bad tuning it might result in some unexpected behavior.

One example is that we saw the velocity of the motor ramp to max velocity when we commanded a setpoint of 10 with the following line of code:

``` java
controller.setReference(10, SparkBase.ControlType.kMAXMotionVelocityControl, 0);
```

Using a feedforward of only 0.01 (see code snippet below). The setup on which we tested this was a NEO motor with only a 1:3 planetary gearbox and a shaft connected to it.

![image](https://github.com/user-attachments/assets/6199566f-e9b4-447a-b3f8-8796ad79f023)


``` java
motor = new SparkMax(2, SparkLowLevel.MotorType.kBrushless);
encoder = motor.getEncoder();
controller = motor.getClosedLoopController();

// Configure drive motor
var config = new SparkMaxConfig();
config
    .encoder
    .positionConversionFactor(1.0)
    .velocityConversionFactor(1.0)
    .uvwMeasurementPeriod(10)
    .uvwAverageDepth(2);
config
    .closedLoop
    .feedbackSensor(ClosedLoopConfig.FeedbackSensor.kPrimaryEncoder)
    .pidf(0.0,0.0,0.0,0.01)
    .maxMotion
    .positionMode(MAXMotionConfig.MAXMotionPositionMode.kMAXMotionTrapezoidal)
    .maxAcceleration(1000.0)
    .maxVelocity(1000.0)
    .allowedClosedLoopError(0.001);
config
    .idleMode(SparkBaseConfig.IdleMode.kCoast)
    .smartCurrentLimit(60)
    .voltageCompensation(12.0);
config
    .signals
    .primaryEncoderPositionAlwaysOn(true)
    .primaryEncoderPositionPeriodMs(20)
    .primaryEncoderVelocityAlwaysOn(true)
    .primaryEncoderVelocityPeriodMs(20)
    .appliedOutputPeriodMs(20)
    .busVoltagePeriodMs(20)
    .outputCurrentPeriodMs(20);
motor.configure(config, SparkBase.ResetMode.kResetSafeParameters, SparkBase.PersistMode.kPersistParameters);
encoder.setPosition(0.0);
```

**Reply by JochemvdBroek** [reply]
Would this be better to turn into an github issue or is it fine to keep the discussion about Max Motion in here?

**Comment by narmstro2020** [comment]
How do you use the simulator SparkSim.  

The iterate function requires a velocity which you get from DCMotorSim. But DCMotor sim requires an input voltage.  Where do you get the input voltage from.  Is that accessible through the spark device you are simulating???

**Reply by narmstro2020** [reply]
My next question is.  does the iterate method even work.  I tried to command a velocity setpoint of 1000 rpm and case 1 in the iterate method kept reporting 0.0 applied output in the loop.  The input velocity started at 0.0 because that's the initial state of the simulator and the output kept returning 0. 

The pid values were 0, 0, 0, and FF value was 1.0 / RPM of motor

**Reply by narmstro2020** [reply]
Confirmed case 1 in the iterate block isn't calculating an applied output for any velocity/setpoint combo

**Reply by narmstro2020** [reply]
Confirmed for position as well.

**Reply by narmstro2020** [reply]
I figured a code snippet would be helpful 

`    public TestSubsystem() {
        sparkMax = new SparkMax(10, MotorType.kBrushless);
        double ff = 1.0 / Units.radiansPerSecondToRotationsPerMinute(DCMotor.getNEO(1).freeSpeedRadPerSec);
        ClosedLoopConfig closedLoopConfig = new ClosedLoopConfig().p(0.0).i(0.0).d(0.0).velocityFF(ff);
        
        SparkMaxConfig sparkMaxConfig = new SparkMaxConfig();
        sparkMaxConfig.apply(closedLoopConfig);
        sparkMax.configure(sparkMaxConfig, ResetMode.kNoResetSafeParameters, PersistMode.kNoPersistParameters);
        sparkMaxSim = new SparkMaxSim(sparkMax, DCMotor.getNEO(1));
        sparkMaxSim.enable();
        LinearSystem<N2, N1, N2> plant = LinearSystemId.createDCMotorSystem(DCMotor.getNEO(1), 0.00001, 1);
        dcMotorSim = new DCMotorSim(
                plant,
                DCMotor.getNEO(1));
        setDefaultCommand(run(() -> sparkMax.getClosedLoopController().setReference(1000, ControlType.kVelocity)));
    }

    @Override
    public void simulationPeriodic() {
        double appliedOutput = sparkMaxSim.getAppliedOutput();
        double busVoltage = sparkMaxSim.getBusVoltage();
        double inputVoltage = appliedOutput * busVoltage;
        dcMotorSim.setInput(inputVoltage);
        dcMotorSim.update(0.020);
        double velocityRPM = dcMotorSim.getAngularVelocityRPM();
        sparkMaxSim.iterate(velocityRPM, 12, 0.020);
    }`

**Comment by viggy96** [comment]
Is there a way to permanently disable a status frame, even if a getter is called?
For example, disable analog sensor status frame, even if I call `m_spark.getAnalog().getPosition()` ?

**Reply by jfabellera** [reply]
No there isn't a way to do that. What would the use case be?

**Reply by viggy96** [reply]
I'm creating a wrapper for the Spark Flex/Max that has AdvantageKit logging built-in. So I'm logging the motor encoder position/velocity, absolute encoder position/velocity, analog sensor position/velocity, limit switches, faults, warnings, motor temperature, applied output, basically everything.

With the aforementioned functionality, I could log all these values, but the ones that are irrelevant (the ones that aren't useful for that particular motor) would simply return zero, and those irrelevant status frames wouldn't take up valuable bandwidth on the bus.

**Reply by viggy96** [reply]
Alternatively, is there a way to query what status frames are enabled?

**Comment by viggy96** [comment]
Using the WPILib Units API for all relevant methods would be a great change. It basically eliminates the footgun of unit conversions.

**Reply by PeterJohnson** [reply]
FYI the WPILib direction we're headed for this for 2027 is to use SI doubles as the default, and add unit-taking overloads, rather than going units-only.  The problem is the Java units library is too complicated both usage and syntax wise for middle school students without mentor support to tackle.

**Reply by narmstro2020** [reply]
I remember at one point there being a document for the WPILib roadmap.  Does that still exists and as it been updated.

**Reply by viggy96** [reply]
I guess that's a reasonable compromise in WPILib, something similar in REVLib would be nice. 
Not sure that the modern Units API is that difficult to use in and of itself, but rather dealing with units objects in calculations or methods that don't support them.

**Comment by jfabellera** [comment]
REVLib beta 4, SPARK MAX and Flex pre-release 9, and Hardware Client 1.7.0-frcBeta2025.1 have been released. This may be the last REV beta release until kickoff.

* [REVLib 2025.0.0-beta-4](https://github.com/REVrobotics/REV-Software-Binaries/releases/tag/revlib-2025.0.0-beta-4)
* [Instructions for updating Hardware Client and firmware](https://github.com/wpilibsuite/2025Beta/blob/main/REV.md)

> [!WARNING]
> After updating your SPARK MAX or Flex to 25.0.0, you will be unable to communicate with them through the Hardware Client if using a Power Distribution Hub or Pneumatic Hub as the USB CAN bridge. A later update will be available for the PDH and PH that fixes this. For now, connect to the Hardware Client via a SPARK MAX or Flex.

### Issues fixed
- #77
- #73
- #66
- #53
- #31

**Comment by narmstro2020** [comment]
So any word on whether or not actual closed loop control (using ks, kv, ka, kg, pid and FF type (elev, arm)) will be implemented.  

I know arbFF is a thing, but it would still require doing calculations on the RIO.

**Reply by viggy96** [reply]
It's kinda required when using motion profiling (like MAXMotion) since the feed forward value is dependent on the updating setpoint, so the ESC needs to calculate the feed forward value on its own.


That said, there can be mechanisms where even the lack of internal feed forward calculations end up not mattering, and the behavior in the real world ends up being "good enough".

**Reply by narmstro2020** [reply]
ESC?

**Reply by viggy96** [reply]
Electronic Speed Controller

**Reply by narmstro2020** [reply]
I understand as well, it just would be nice to have the option.

**Reply by viggy96** [reply]
I agree, from a technical perspective it's required like I said, and banking on things just working out without it for all mechanisms for motion profiling isn't really a good strategy.

**Reply by narmstro2020** [reply]
Ideally I would like to get all of my Feedforward and Feedback values (even if they have to be tweaked a bit later from sysId).  

Well I can get those values for ks, kv, ka, pid, etc but I can't use have of them on the Spark directly.

**Comment by Liam-Stow** [comment]
How is kF used in MaxMotion position control calculations? I would have expected it to apply an output equal to the current velocity target * kF.  
My experiments in the REV Hardware client seem to indicate that is not the case, and instead it could be applying an output equal to the current position target (not the overall reference, but just the next point in the profile) * kF. It is my understanding that this means kF is not useful for MaxMotion position control, since any time your current position is > 0, kF will be applying a positive output regardless of your overall reference or desired direction of movement, and anytime your current position is < 0, kF will apply a negative output. Is that correct and thus we should only be using feedback (PID) control for MaxMotion?

My experiment was as follows:
1. Set kF to 0.005, max velocity to 300 and max acceleration to 500. All other control parameters are zero
2. Set a MaxMotion position target of zero
3. Move motor by hand to make the position signal read +10 rotations
4. Enable the motor in REV Hardware client
5. Observe the motor move in the positive direction (away from the target) at an increasing rate
6. Disable the motor in REV Hardware client
7. Move the motor by hand to make the position signal read -10 rotations
8. Enable the motor in REV Hardware client
9. Observe the motor move in the negative direction (away from the target) at an increasing rate
